File: frontend/src/services/api/reportService.ts
import { axiosInstance } from "@/services/api/axiosInstance";
import type {
  DeliveryReportResponse,
  InventoryReportResponse,
  OrdersReportResponse,
  PaymentReportResponse,
  SalesReportResponse,
} from "@/types/report.types";

/**
 * Report API client. All endpoints are SA/OWNER/MANAGER only
 * (SecurityConfig); values come from the real backend.
 *
 * Backend reference: ReportController at /api/reports.
 * `from`/`to` are ISO dates (YYYY-MM-DD); omit for all time.
 */
interface ReportRange {
  from?: string;
  to?: string;
}

export const reportService = {
  async getSalesReport(range?: ReportRange): Promise<SalesReportResponse> {
    const response = await axiosInstance.get<SalesReportResponse>(
      "/reports/sales",
      { params: range },
    );
    return response.data;
  },

  async getOrdersReport(range?: ReportRange): Promise<OrdersReportResponse> {
    const response = await axiosInstance.get<OrdersReportResponse>(
      "/reports/orders",
      { params: range },
    );
    return response.data;
  },

  async getInventoryReport(): Promise<InventoryReportResponse> {
    const response = await axiosInstance.get<InventoryReportResponse>(
      "/reports/inventory",
    );
    return response.data;
  },

  async getDeliveryReport(range?: ReportRange): Promise<DeliveryReportResponse> {
    const response = await axiosInstance.get<DeliveryReportResponse>(
      "/reports/deliveries",
      { params: range },
    );
    return response.data;
  },

  async getPaymentReport(range?: ReportRange): Promise<PaymentReportResponse> {
    const response = await axiosInstance.get<PaymentReportResponse>(
      "/reports/payments",
      { params: range },
    );
    return response.data;
  },
};

File: backend/src/main/java/com/distributrack/service/ReportService.java
package com.distributrack.service;

import com.distributrack.dto.response.DeliveryReportResponse;
import com.distributrack.dto.response.InventoryReportResponse;
import com.distributrack.dto.response.OrdersReportResponse;
import com.distributrack.dto.response.PaymentReportResponse;
import com.distributrack.dto.response.SalesReportResponse;

import java.time.LocalDate;

/**
 * Operational reports. All endpoints are restricted to business roles
 * (SUPER_ADMIN / OWNER / MANAGER) by SecurityConfig. Revenue follows the
 * established rule: only DELIVERED/COMPLETED orders count, once each.
 */
public interface ReportService {

    // Sales Report (rows + summary; optional date range)
    SalesReportResponse getSalesReport(LocalDate from, LocalDate to);

    // Orders Report (rows + lifecycle counts; optional date range)
    OrdersReportResponse getOrdersReport(LocalDate from, LocalDate to);

    // Inventory Report (rows + summary)
    InventoryReportResponse getInventoryReport();

    // Delivery Report (rows + status counts; optional date range)
    DeliveryReportResponse getDeliveryReport(LocalDate from, LocalDate to);

    // Payment Report (rows + totals; optional date range)
    PaymentReportResponse getPaymentReport(LocalDate from, LocalDate to);
}